### PR TITLE
323.coll trailing slash

### PR DESCRIPTION
**Pull request by d-w-moore**
Allow session.collections.get(Cpath) and session.data_objects.put(Dfile,Cpath) to work even when Cpath is not slash-normalized, or has trailing slash(es).

### CODE REVIEW

**Comment by trel** [comment]
i'm not clear on what this is exactly doing.

please add a test or two that shows the cases this handles cleanly.

**Comment by d-w-moore** [comment]
Sure.  Also, ... still seeking out the other corner cases.

File: irods/collection.py [hunk 1]
```diff
@@ -78,5 +85,14 @@ def walk(self, topdown=True):
         if not topdown:
             yield (self, self.subcollections, self.data_objects)
 
+    @staticmethod
+    def normalize_path(path):
+        path_ = _minimize_slashes(path).rstrip('/')
```
Comment (trel) [inline irods/collection.py:90]: should this also handle... dot (`.`) and dot-dot (`..`)?  probably should be a separate issue if so.

@korydraughn do we already have a way to do this in the server - either that can be consulted / called / emulated?

Comment (trel) [reply irods/collection.py:90]: oh... i see something similar being tested below... ignore if this is already handled.

File: irods/test/collection_test.py [hunk 2]
```diff
@@ -5,12 +5,14 @@
...
 import shutil
 import unittest
+import time
 from irods.meta import iRODSMetaCollection
 from irods.exception import CollectionDoesNotExist
 from irods.models import Collection, DataObject
 import irods.test.helpers as helpers
 import irods.keywords as kw
 from six.moves import range
+from irods.test.helpers import my_function_name,unique_name
```
Comment (korydraughn) [inline irods/test/collection_test.py:15]: space after comma.

there are a few places that need this treatment below as well.

File: irods/test/collection_test.py [hunk 3]
```diff
@@ -322,6 +324,33 @@ def test_register_collection_with_checksums(self):
...
+        subcoll, dataobj = [unique_name(my_function_name(),time.time()) for x in range(2)]
+        subcoll_fullpath = "{}/{}".format(Home,subcoll)
+        subcoll_unnormalized = subcoll_fullpath + "/"
+        try:
+            # issue 323  - test create and exists with trailing slashes
+            self.sess.collections.create(subcoll_unnormalized)
+            c1 = self.sess.collections.get( subcoll_unnormalized )
+            c2 = self.sess.collections.get( subcoll_fullpath )
+            self.assertEqual(c1.id, c2.id)
+            self.assertTrue( self.sess.collections.exists(subcoll_unnormalized))
```
Comment (korydraughn) [inline irods/test/collection_test.py:338]: inconsistent spacing around function arguments.

there are other places that need to be cleaned up as well.

**Review by korydraughn — COMMENTED** [review]
looks good overall.

File: irods/collection.py (hunk 1 above)
Comment (d-w-moore) [reply irods/collection.py]: I think I've handled the ` . ` and ` .. ` now, but will write a test for that too.

File: irods/test/collection_test.py (hunk 3 above)
Comment (d-w-moore) [reply irods/test/collection_test.py]: Think these are fine now.

File: irods/collection.py (hunk 1 above)
Comment (d-w-moore) [reply irods/collection.py]: I could still make `.` and `..` a separate issue.  (Related: session `chdir`, `cwd`,  and other "enhancements"  that are actually extras. (The new `iRODSPath` class is not extraneous - it performs the path normalization in direct support of this issue, #323)

Comment (trel) [reply irods/collection.py]: yeah, maybe . and .. should be separately described/handled in a new issue.

File: irods/collection.py [hunk 4]
```diff
@@ -78,5 +84,20 @@ def walk(self, topdown=True):
         if not topdown:
             yield (self, self.subcollections, self.data_objects)
 
+    @staticmethod
+    def normalize_path(*paths_, **kw_):
+        """Normalize a path or list of paths.
+
+        We use the iRODSPath class to eliminate extra slashes in,
+        and (if more than one parameter is given) concatenate, paths.
+        if keywords include enforce_absolute, this function requires
```
Comment (trel) [inline irods/collection.py:93]: "if" -> "If"

File: irods/test/collection_test.py [hunk 5]
```diff
@@ -322,6 +325,52 @@ def test_register_collection_with_checksums(self):
...
+            normalize('../public')
+
+        # Test '.' and double slashes
+        public_home = normalize(home,'..//public/.//')
+        self.assertEqual(public_home, '/{sess.zone}/home/public'.format(sess = session))
+
+        # Test that '..' cancels last nontrival path element
+        subpath = normalize(home,'./collA/coll2/./../collB')
+        self.assertEqual(subpath, home + "/collA/collB")
+
```
Comment (trel) [inline irods/test/collection_test.py:372]: please add two more cases where there are multiple '..' somewhere in the path...
 - both right next to one another 
 - spread out by a couple elements

File: irods/path/__init__.py [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+        s = '/'.join(list_of_str)
+        if s[:1] == '/':
+            absolute = True
+        components = iRODSPath.resolve_noops_and_updirs( _multiple_slash.split(s), absolute)
+        path = (
+                '/' if absolute else ''
+               ) + (
+                '/'.join(filter(None, components))
+               )
+        return path if path else '.'
```
[the remark this answers is not in the document]
Comment (d-w-moore) [reply irods/path/__init__.py:68]: I don't think that it does enough of the job - for one, it doesn't remove `.` and `..` as would be required for any path being submitted as a logical path to iRODS.

[the remark this answers is not in the document]
Comment (alanking) [reply irods/path/__init__.py:68]: Ah, okay. I figured this might be a naive take, just figured I'd throw it out there

[the remark this answers is not in the document]
Comment (d-w-moore) [reply irods/path/__init__.py:68]: Possibly a `lambda x: os.path.normpath(os.path.join('/',x))
`

[the remark this answers is not in the document]
Comment (d-w-moore) [reply irods/path/__init__.py:68]: Better still` lambda x,abs=True:os.path.normpath(os.path.join(('/' if abs else ''),x))
`

File: irods/test/collection_test.py (hunk 5 above)
Comment (d-w-moore) [reply irods/test/collection_test.py:376]: done.

File: irods/collection.py (hunk 4 above)
Comment (d-w-moore) [reply irods/collection.py]: Done....

File: irods/path/__init__.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (d-w-moore) [reply irods/path/__init__.py]: This is improved now, much smaller code, and no longer the "custom-made" kind.

File: irods/collection.py (hunk 1 above)
Comment (d-w-moore) [reply irods/collection.py]: #344 created.

**Comment by trel** [comment]
after that, please squash to a single(?) commit?

**Comment by d-w-moore** [comment]
> after that, please squash to a single(?) commit?
ok

File: irods/path/__init__.py [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+
+    def __new__(cls, *elem_list, **kw):
+        """
+        Initializes our immutable 'str' object with a normalized form.
+
+        Keywords may include 'absolute'. The default is True, forcing a slash as
+        the leading character of the resulting string object.
+
+        Variadic parameters are the path elements, strings which may name individual
+        collections or sub-hierarchies (internally slash-separated):
```
[the remark this answers is not in the document]
Comment (d-w-moore) [reply irods/path/__init__.py]: Ok added in this example.  See if you think it scans ok?  Then I'll # and squash.

File: irods/path/__init__.py [hunk 8]
```diff
@@ -9,20 +9,39 @@ class iRODSPath(str):
...
 
         Variadic parameters are the path elements, strings which may name individual
-        collections or sub-hierarchies (internally slash-separated):
+        collections or sub-hierarchies (internally slash-separated).  These are then
+        joined using the path separator:
 
-        In the resulting object, trailing and redundant path separator are removed,
-        as is the "trivial" path element ('.'):
+          data_path = iRODSPath( 'myZone', 'home/user', './dir', 'mydata')
+          # => '/myZone/home/user/dir/mydata'
```
Comment (trel) [inline irods/path/__init__.py:23]: yeps.

indent a couple more spaces to match the ones below...

File: irods/path/__init__.py [hunk 9]
```diff
@@ -9,20 +9,39 @@ class iRODSPath(str):
...
+        of the `absolute' hint, but subsequent arguments will act as relative paths
+        regardless of leading slashes. So this will do the "right thing":
+
+            my_dir = str(iRODSPath('dir1'))                     # => "/dir1"
+            my_rel = ""+iRODSPath('dir2', absolute=False)       # => "dir2"
+            my_abs = iRODSPath('/Z/home/user', my_dir, my_rel)  # => "/Z/home/user/dir1/dir2"
+
+        Finally, because iRODSPath has class `str` as a base class, this is also possible:
+
+            iRODSPath( '//zone/home/public/this', iRODSPath('../that',absolute=False))
```
Comment (trel) [inline irods/path/__init__.py:48]: please show the result, just in case someone can't 'see' it.

File: irods/path/__init__.py (hunk 8 above)
Comment (d-w-moore) [reply irods/path/__init__.py]: done.

File: irods/path/__init__.py (hunk 9 above)
Comment (d-w-moore) [reply irods/path/__init__.py]: Done...

**Comment by trel** [comment]
typo in your commit message - "underlyiing"

fix that and we'll merge it.

**Comment by d-w-moore** [comment]
ok, done!

**Review by alanking — APPROVED** [review]
